### PR TITLE
Add div/mod by zero and SWAP stack tests

### PR DESCRIPTION
**Pull request by MrChico**
This PR adds test fillers which test DIV, SDIV, MOD and SMOD by zero as requested by https://github.com/ethereum/tests/issues/459
and a missing stack checks for SWAP that lead to a bug slipping into KEVM https://github.com/kframework/evm-semantics/pull/482

@winsvega I haven't been able to run `retesteth` to generate actual tests from these fillers. I get the error message `A test file must contain an object.` when running
```
retesteth -t GeneralStateTests/stSolidityTest -- --testpath ~/tests
```
and errors about `ByzantiumToConstantinopleFixAt5 not found in correct mining info config!` when running 
```
retesteth -t GeneralStateTests/stSolidityTest -- --fillchain --testpath ~/tests
```

Until documentation improves or error messages becomes more descriptive, I can only provide fillers and not actual tests.

### CODE REVIEW

**Comment by winsvega** [comment]
~~the test generation would work in your case in you remove --fillchain.~~
~~I forgot to update mining reward config for new forks thats why you see the error.~~ 
~~however it shows me that transaction in your test is configured as invalid.~~

~~the filling fails as the transaction is rejected by the client as invalid. state tests must have valid transactions. (enough intrinsic gas, and so on)
for some reason geth dropped the transaction and didn't return it with the block info. maybe this is an issue in geth the transaction in the test looks good~~

I've improved the error checking and report on retesteth for this cases. 
update the test repo and retesteth:develop branch. the more documentation would be great. 
for now I improve the usability by fixing the issues by your's report and making the error info more informative.  

you forgot to put transaction sender account to the pre state. thats why transaction fail to execute on remote clients.

**Comment by MrChico** [comment]
Thanks @winsvega . I was now able to fill the `BlockchainTests`. I am still unsuccessful in filling the `GeneralStateTests`. Running:
```
~/retesteth/build/retesteth/retesteth -t GeneralStateTests/stStackTests -- --testpath ~/tests
```
I get the error message:
```
Error: Contents of 
tests/GeneralStateTests/stStackTests/stackOverflowSWAP.json is empty. Have you filled the test?
Error: Contents of 
tests/GeneralStateTests/stStackTests/stacksanitySWAP.json is empty. Have you filled the test? (stStackTests/stackOverflowM1, fork: Istanbul, TrInfo: d: 15, g: 0, v: 0)
```
My understanding was that this is how I'm supposed to fill the tests. Do you know what the problem might be?

**Comment by winsvega** [comment]
To fill the test stackOverflowSWAP

`./retesteth -t GeneralStateTests/stStackTests -- --singletest stackOverflowSWAP --filltests`

**Comment by MrChico** [comment]
Ah, which is already present in the documentation. Thank you @winsvega this should now be ready for review.

File: src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            }
+        },
+        "//comment" : "multiple transaction description that will be applied on a pre state",
+        "transaction" : {
+            "//comment" : "foreach d in `data` { foreach g in `gasLimit` { foreach v in `value`  {  pre.applyTransaction(d, g, v)  }}}",
+            "//comment" : "each element of the data array is treated same as account's code field. You could use LLL here",
+            "data" : [
+              "0x6001600004600055",
+              "0x6001600005600055",
+              "0x6001600006600055",
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: In expect section you should check the result of 55 sstore operation in a newly created account.

There must be a way to check that test has worked or not. If it changes 0 to 0 there is no way to detect that it has worked. (Other then checking the gas usage which is not preferable)

Comment (MrChico) [reply src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: Thanks. Addressed this in the new commit

Comment (winsvega) [reply src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: btw https://github.com/ethereum/tests/blob/develop/PRLOG.md

File: src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json [hunk 2]
```diff
@@ -23,16 +23,16 @@
                 "network" : [">=Istanbul"],
                 "result" : {
                     "//comment" : "Account in expect section would be checked for the fields specified here (balance, code, storage)",
-                    "095e7baea6a6c7c4c2dfeb977efac326af552d87" : {
-                        "balance" : "1000000000000000000",
+                    "6295ee1b4f6dd65047762f924ecd367c17eabf8f" : {
+                        "balance" : "0",
                         "//comment" : "No code is returned",
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json:25]: I think this template comments are irrelevant

File: src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json [hunk 3]
```diff
@@ -23,16 +23,16 @@
...
                         "code" : "",
                         "storage" : {
-                            "0x00" : "0x00"
+                            "0x00" : "0x01"
                         }
                     },
                     "a94f5374fce5edbc8e2a8697c15331677e6ebf0b" : {
-                      "balance" : "999999999999946082",
+                      "balance" : "999999999999926870",
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: balance check is not needed. the test seems to work fine with any gasCost rescheduled

File: src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json [hunk 4]
```diff
@@ -63,13 +53,12 @@
...
         "transaction" : {
-            "//comment" : "foreach d in `data` { foreach g in `gasLimit` { foreach v in `value`  {  pre.applyTransaction(d, g, v)  }}}",
-            "//comment" : "each element of the data array is treated same as account's code field. You could use LLL here",
+            "//comment" : "store 1 at the result of {DIV,SDIV,MOD,SMOD}(1,0)",
             "data" : [
-              "0x6001600004600055",
-              "0x6001600005600055",
-              "0x6001600006600055",
-              "0x6001600007600055"
+              "0x6001600160000455",
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: I get this code from lllc  "[[0]] (div 1 0)" -> "600060010460005500"

Comment (winsvega) [reply src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: you could also use LLLC in transaction data

Comment (MrChico) [reply src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: > I get this code from lllc "[[0]] (div 1 0)" -> "600060010460005500"

That has different behaviour. I am storing the number 1 at storage location `div(1,0)`

Comment (winsvega) [reply src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: lllc "[[ (div 1 0) ]] 1" -> "600160006001045500"  is also different
maybe just to be safe you could use lllc in transaction data directly?

also have you put a record into https://github.com/ethereum/tests/blob/develop/PRLOG.md ?

**Comment by MrChico** [comment]
Thank you for the review @winsvega , I fixed the bytecode again and added a PRLOG entry

File: src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json [hunk 5]
```diff
@@ -32,7 +29,7 @@
                         }
                     },
                     "a94f5374fce5edbc8e2a8697c15331677e6ebf0b" : {
-                      "balance" : "999999999999926870",
+                      "balance" : "999999999999926866",
```
Comment (winsvega) [inline src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: please remove this balance expect

Comment (MrChico) [reply src/GeneralStateTestsFiller/stSolidityTest/ByZeroFiller.json]: fixed

**Comment by holiman** [comment]
Did this contain any mulmod with `(a,b,0)` by any chance?

**Comment by MrChico** [comment]
No, only `DIV`, `SDIV`, `MOD` and `SMOD` went in here I'm afraid

**Pull request merged**